### PR TITLE
Add evaluation framework

### PR DESCRIPTION
**Pull request by ink-pad**
# Add Granite Guardian evaluation framework

Adds `evaluation/`, a minimal framework to reproduce the benchmark numbers on the Granite Guardian model cards for the whole family (3.0 through 4.1). It loads a Guardian model with vLLM, runs one or more benchmarks, and writes per-dataset metrics plus an aggregate.

## What's included

Four benchmarks, matching what the cards report:

| Benchmark | Measures | Metric | Data |
|---|---|---|---|
| `ood_safety` | Harm / out-of-distribution safety (10 datasets) | F1 | local prep |
| `groundedness_true` | RAG hallucination on TRUE (9 datasets) | AUC / BAcc | local prep |
| `groundedness_aggrefact` | RAG hallucination on LM-AggreFact | balanced accuracy | turnkey (HF) |
| `function_calling` | Function-call hallucination on FC-Reward-Bench | balanced accuracy | turnkey (HF) |

All eight released models are supported through three prompt-API modules (`granite-guardian-3` for 3.0-3.2, `granite-guardian-3.3`, `granite-guardian-4.1`); `--model-type` is auto-inferred from the model path. Both no-think and `--think` modes are supported for 3.3 and 4.1.

The two turnkey benchmarks pull their data from the Hugging Face Hub and need no setup. The harm and TRUE sets are curated subsets we do not redistribute; the repo ships the loader plus sourcing instructions (`data/README.md`), not data.

## Replication of the card numbers

Every model module was verified end-to-end by running this exact code against the model cards on GPU. Results reproduce the published numbers to within run-to-run variance:

| Model | Metric | This framework | Model card |
|---|---|---|---|
| 4.1-8b | Harm F1 | 0.7945 | 0.79 |
| 4.1-8b | LM-AggreFact BAcc | 0.760 | 0.760 |
| 4.1-8b | FC-Reward-Bench BAcc | 0.7903 | 0.79 |
| 3.3-8b | Harm F1 | 0.8104 | 0.81 |
| 3.3-8b (no-think) | LM-AggreFact BAcc | 0.758 | 0.761 |
| 3.3-8b (think) | LM-AggreFact BAcc | 0.755 | 0.765 |
| 3.3-8b | FC-Reward-Bench BAcc | 0.742 | 0.74 |
| 3.0-8b | Harm F1 | 0.7643 | 0.76 |
| 3.0-8b | TRUE AUC | 0.8405 | 0.85 |

One documented caveat: the TRUE **balanced-accuracy** operating point is fixed at 0.5, which reads slightly below the 3.3 card (~0.76 vs 0.777). The document ranking is unaffected (TRUE AUC ~0.873), so this is an operating-point difference rather than a modelling one; use the AUC columns for a threshold-free comparison. This is called out in the README.

### Source of the 3.3 LM-AggreFact number

The **0.765** LM-AggreFact figure on the Granite Guardian 3.3 card is produced by the official MiniCheck project ([Liyan06/MiniCheck](https://github.com/Liyan06/MiniCheck), Apache-2.0), which ships a first-class `Granite-Guardian-3.3-8B` fact-checking path and reports it on the [LLM-AggreFact leaderboard](https://llm-aggrefact.github.io/). That path runs the model in think mode with a SentenceFusion protocol (the document is chunked, the claim is split into sentences, the score is the max over chunks then the min over sentences). Running MiniCheck's code unmodified against the same data reproduces the card: **avg-per-dataset BAcc 0.763** (vs card 0.765).

This framework ships a single-pass groundedness path instead (one guard call per document/claim pair), which is exactly what the 4.1 pipeline uses. On 3.3 it lands at 0.758 (no-think) / 0.755 (think) - within a point of the card, matching MiniCheck
to <1 point on 10 of 11 datasets, with the whole gap coming from AggreFact-CNN (the one multi-sentence-summary set, where SentenceFusion's per-sentence handling differs). No MiniCheck code or data is vendored into this PR. Anyone wanting the exact card number should cite and run MiniCheck; this framework's number is a close, self-contained reproduction with the protocol difference documented. The 4.1 LM-AggreFact card value (0.760) is reproduced exactly by this framework's own single-pass path.

## Reproducing

```bash
pip install -r evaluation/requirements.txt

# Turnkey (no data setup): LM-AggreFact + FC-Reward-Bench on 4.1
python evaluation/run_eval.py \
    --model-path ibm-granite/granite-guardian-4.1-8b \
    --benchmarks groundedness_aggrefact function_calling

# Harm + TRUE require locally-prepared data (see evaluation/data/README.md)
export GG_EVALS_DATA_ROOT=/path/to/your/eval_data
python evaluation/run_eval.py --model-path ibm-granite/granite-guardian-3.3-8b
```

Add `--think` for reasoning mode (3.3 / 4.1). `--num-shards N` data-parallelizes the shardable benchmarks across GPUs.

## Extensible

Models, benchmarks, and data are three independent drop-in registries. A new guard (e.g. Llama Guard, ShieldGemma) is a single `models/<name>.py` exporting `format_fn` / `parse_fn` / `MODEL_NAME`; it is auto-discovered as a new  `--model-type` with no other changes. See the "Bring your own ..." section of the README.

## Notes for reviewers

- No datasets are committed; the two local-prep benchmarks read from a user-provided `GG_EVALS_DATA_ROOT`.
- Some HF datasets are gated (accept terms on the Hub and set `HF_TOKEN`).

### CODE REVIEW

File: evaluation/benchmarks/groundedness_true.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+under GG_EVALS_DATA_ROOT.
+"""
+
+import os
+import pandas as pd
+from datasets import Dataset
+from tqdm import tqdm
+
+import nltk
+nltk.download("punkt_tab", quiet=True)
```
Comment (mnagired) [inline evaluation/benchmarks/groundedness_true.py]: this is called at module import and is possible that this fails (or silently yields no
tokenizer) — worth documenting as a pre-flight step, or wrapping so the failure
message is actionable.

File: evaluation/models/granite_guardian_4_1.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    label = None
+    match = re.findall(r"<score>\s*(.*?)\s*</score>", text_for_parse, re.DOTALL)
+    if match:
+        score = match[0].strip().lower()
+        if "yes" in score:
+            label = 1
+        elif "no" in score:
+            label = 0
+
+    if label is None:  # fallback: bare yes/no anywhere in the (think-stripped) text
```
Comment (mnagired) [inline evaluation/models/granite_guardian_4_1.py]: falls back to a bare yes/no scan when the `<score>` tags are missing; `models/granite_guardian_3_3.py` has no such fallback, so a 3.3 output missing the tags is counted as a parse failure (and then scored as the wrong label by `compute_metrics`). Not necessarily wrong — just an undocumented
asymmetry that could move the 3.3 numbers relative to 4.1.

File: evaluation/models/_helpers.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
+"""Shared helpers for Granite Guardian model modules."""
+
+import json
+import math
+
+import torch
```
Comment (mnagired) [inline evaluation/models/_helpers.py]: imports `torch` solely to compute a softmax over two scalars
(`softmax_yes_no` / `softmax_safe_unsafe`). `math.exp` would do, dropping a heavy
import for anyone reusing these helpers outside a GPU context. `torch` is already a
transitive dep of vllm, so this is cosmetic.

File: evaluation/benchmarks/groundedness_aggrefact.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+"""
+
+import json
+import os
+
+import pandas as pd
+from datasets import load_dataset
+from sklearn.metrics import balanced_accuracy_score
+
+HF_DATASET = "lytang/LLM-AggreFact"
```
Comment (mnagired) [inline evaluation/benchmarks/groundedness_aggrefact.py:19]: calls `load_dataset("lytang/LLM-AggreFact", split="test")` with no `revision=`. LM-AggreFact is a living dataset (the leaderboard adds datasets over time), so the reproduced number silently depends on whatever revision the user happens to fetch. Pinning a `revision=` (or printing the resolved
commit hash) would make the 0.760 / 0.758 claims reproducible over time. 

Same applies to `ibm-research/fc-reward-bench`.

**Review by mnagired — CHANGES_REQUESTED** [review]
minor changes, otherwise looks good! thanks a lot @ink-pad!

File: evaluation/models/_helpers.py (hunk 3 above)
Comment (ink-pad) [reply evaluation/models/_helpers.py]: Resolved.

**Comment by ink-pad** [comment]
@mnagired Thanks a lot. The last commit should have resolved them all.

**Review by mnagired — APPROVED** [review]
LGTM, thanks!
